android/app/src/main/java/org/fossasia/openevent/core/about/AboutFragmentViewModel.java

```diff
@@ -58,10 +59,12 @@ public AboutFragmentViewModel() {
     }
 
     public LiveData<List<Object>> getBookmarkedSessions() {
-        if (sessions == null) {
+        if (sessions == null || numberOfBookmarkedSessions != realmRepo.getBookMarkedSessionsLength()) {
             LiveRealmData<Session> sessionLiveRealmData = RealmDataRepository.asLiveData(realmRepo.getBookMarkedSessions());
             sessions = Transformations.map(sessionLiveRealmData, this::getSessionsList);
```

Comment:
`realmRepo.getBookMarkedSessions()` these should be reactive and live and hence should be updated automatically

Reply:
But without going inside the if condition how the sessions object will be updated because we are returning sessions object. I added this condition "numberOfBookmarkedSessions != realmRepo.getBookMarkedSessionsLength()" so that the body of the if condition will be executed only when there is change in the realm.

Reply:
Actually I tried and changed the body of the if condition to this,

``` 
 if(sessions == null) {
      sessions = Transformations.map(RealmDataRepository.asLiveData(realmRepo.getBookMarkedSessions()),this::getSessionsList);
   }  
```

I changed this because realmRepo.getBookMarkedSessions() should be executed if the object is auto updated but still after changing the code to this sessions object returns the old list.

Reply:
I would recommend that you actually try to get bookmarked sessions from RealmRepository and add a listener on them and check if you are getting change event or not. This will reveal the problem

Reply:
Ok.

Reply:
> Also I think that sessions object is an object of LiveData so why it should be auto updated.

This is exactly why it should get updated. LiveData is meant to be reactive and change aware. So if it is not being updated, it is a bug in the implementation of conversion into LiveData

As to why your code is not working, you create an instance of `sessionLiveRealmData`, let's say it is instance `ada2313v`, so the fragment is listening to `ada2313v` and at each change you just reassign the reference to `sessionLiveRealmData` to a new instance, let's say `e4ew335r`. The fragment is still listening to `ada2313v`, it doesn't care about your local change. If things were simple (reactive) this way, we wouldn't even need LiveData, Observables and everything. Want to notify an object has changed, just reassign the variable to a new thing and it will be changed everywhere!

This confusion is the reason why we should always use final (non-reassignable) properties wherever possible so that changes can be propagated by mutating the LiveData and not by reassigning it (which does nothing but make the app slower by Object Instantiation)

Reply:
``` So if it is not being updated, it is a bug in the implementation of conversion into LiveData ``` 
Yes, it is the bug in RealmDataRepository.asLiveData(sessions); this method is not returning a new LiveData object. So should I make changes in this method ?

Reply:
Yup

Reply:
@iamareebjamal I got one solution and it is working. I changed the parent class of LiveRealData.java from LiveData<RealmResults<T>> to MutableLiveData<RealmResults<T>> 
and changed the constructor to this

```
public AboutFragmentViewModel() {
        realmRepo = RealmDataRepository.getDefaultInstance();
        s = realmRepo.getBookMarkedSessions();
        sessionLiveRealmData = RealmDataRepository.asLiveData(s);
        s.addChangeListener(sessions -> {
            sessionLiveRealmData.setValue(sessions);
        });
    }
```
The above solution is working. You were right, the ```Transformations.map(sessionLiveRealmData, this::getSessionsList) ``` method was listening to old reference of ```sessionLiveRealmData``` so it will not listen to new data and so I changed the  ```sessionLiveRealmData``` object parent class and changed it's value inside the listener. And it is working fine. Can you tell if anything wrong in this method ?

Reply:
Well obviously this will work because now we are manually triggering the change through a change listener. The idea of RealmLiveData is to automatically do it removing the need of manual changes. This is not the solution unfortunately. I am currently busy or would have seen what's the issue. @srv-twry Can you please help?

Reply:
@iamareebjamal  I thought of this solution as there is no way to update the value of LiveData object and the Transformations.map() method only executes if there is a change in the LIveData object passed in it as a parameter and since it was listening to old LiveData object(i.e. sessionLiveRealmData ) there is no other way to update it and the documentation says that MutableLiveData has the setValue method to change the stored value. That's why I thought of changing the parent class as the only difference between LiveData and MutableLiveData is that MutableLiveData has two extra methods to change the stored value. And that's why I thought of this solution. But if you say so I'll try to think of something else.

Reply:
@iamareebjamal Yes, looking into it.

Reply:
@raj10071997 After going through the issue, i don't think we need to think so much about it. Everything was correctly implemented except in the AboutFragment's ```loadData``` method.

Can you please replace 
``` Java
aboutFragmentViewModel.getBookmarkedSessions().observe(this, sessionsList -> {
            sessions.clear();
            sessions.addAll(sessionsList);
            bookMarksListAdapter = new GlobalSearchAdapter(sessions, getContext());
            bookMarksListAdapter.setOnBookmarkSelectedListener(this);
            bookmarksRecyclerView.setAdapter(bookMarksListAdapter);
            handleVisibility();
        });
```

with this 

``` Java
aboutFragmentViewModel.getBookmarkedSessions().observe(this, sessionsList -> {
            sessions.clear();
            sessions.addAll(sessionsList);
            bookMarksListAdapter.notifyDataSetChanged();
            handleVisibility();
        });
```

Reply:
![videotogif_2018 03 07_01 11 59](https://user-images.githubusercontent.com/21277837/37054406-cc56ab36-21a4-11e8-9dc6-56ab02b84750.gif)

Reply:
@srv-twry I tried your changes locally. I only made the changes you specified and tested them locally. In my phone the bookmarked sessions appear only when they are equal to 3. And if I try to mark more than that, then only 4 bookmarked sessions are shown and then when the sessions become equal to 6, all the bookmarked sessions are shown. The gif shows only the issue about the 3 sessions.

![videotogif_2018 03 07_13 22 49](https://user-images.githubusercontent.com/24502136/37080293-bd0ad7a8-220b-11e8-97f5-d12678058535.gif)

Reply:
@srv-twry The changes you mentioned are the only changes that needed to be done?

Reply:
Yes, they are the only changes that i made. 
The error you're getting is weird. Can you completely uninstall the application and reinstall it after making the above changes ?

Reply:
I am still getting the same results. I also noticed one thing that when you install the app so you are basically logged out (now don't login) and then if you try to bookmark the events, the events don't get displayed at all in AboutFragment.java, doesn't matter how many events you bookmark. Are you getting the same results too when you install the app and you are logged out? I am using Android 7.0 Nougat (API 24).

Reply:
The above GIF that i uploaded has me logged out only.

Reply:
Ok. I'll try to see why am I getting these errors. As I cleared the cache data and uninstalled the application and then installed it with your suggested changes. I have my development branch updated with the remote branch. And still I am getting these errors. I'll see why am I getting these errors and will notify you soon.